kernel.c

```diff
@@ -0,0 +1,728 @@
+          
+        }
+    
+      } 
+      return -1;
+  
+    }
+    
+    proc->region1_pt[i].pfn = index;
+    proc->region1_pt[i].valid = 1;
+    proc->region1_pt[i].prot = PROT_READ | PROT_WRITE;
+
+  }
+
+
+  /*
+   * ==>> (Finally, make sure that there are no stale region1 mappings left in the TLB!)
+   */
+
+  // 1) Flush any old region-1 mappings
+  WriteRegister(REG_TLB_FLUSH, TLB_FLUSH_1);
+
+  // 2) Tell the MMU to use THIS process page table
+  WriteRegister(REG_PTBR1, (unsigned int)proc->region1_pt);
+  WriteRegister(REG_PTLR1, MAX_PT_LEN);
+
+  // 3) Flush again to clear any stale entries under the new table
+  WriteRegister(REG_TLB_FLUSH, TLB_FLUSH_1);   WriteRegister(REG_TLB_FLUSH, TLB_FLUSH_1);
+
+  /*
+   * All pages for the new address space are now in the page table.
+   */
+
+  /*
+   * Read the text from the file into memory.
+   */
+  lseek(fd, li.t_faddr, SEEK_SET);
+  segment_size = li.t_npg << PAGESHIFT;
+  if (read(fd, (void *) li.t_vaddr, segment_size) != segment_size) {
+    close(fd);
+    return KILL;   // see ykernel.h
+  }
+
+  /*
+   * Read the data from the file into memory.
+   */
+  lseek(fd, li.id_faddr, 0);
+  segment_size = li.id_npg << PAGESHIFT;
+
+  if (read(fd, (void *) li.id_vaddr, segment_size) != segment_size) {
+    close(fd);
+    return KILL;
+  }
+
+
+  close(fd);                    /* we've read it all now */
+
+
+  /*
+   * ==>> Above, you mapped the text pages as writable, so this code could write
+   * ==>> the new text there.
+   *
+   * ==>> But now, you need to change the protections so that the machine can execute
+   * ==>> the text.
+   *
+   * ==>> For each text page in region1, change the protection to (PROT_READ | PROT_EXEC).
+   * ==>> If any of these page table entries is also in the TLB,
+   * ==>> you will need to flush the old mapping.
+   */
+
+
+  for (int i = text_pg1; i < text_top; i++){
+
+    proc->region1_pt[i].prot = PROT_READ | PROT_EXEC;
+
+  }
+
+  WriteRegister(REG_TLB_FLUSH, TLB_FLUSH_1);
+
+
+  /*
+   * Zero out the uninitialized data area
+   */
+  bzero((void*)li.id_end, li.ud_end - li.id_end);
+
+  /*
+   * Set the entry point in the process's UserContext
+   */
+
+  /*
+   * ==>> (rewrite the line below to match your actual data structure)
+   * ==>> proc->uc.pc = (caddr_t) li.entry;
+   */
+
+  proc->uctxt.pc = (caddr_t)li.entry;
+
+  /*
+   * Now, finally, build the argument list on the new stack.
+   */
+
+
+  memset(cpp, 0x00, VMEM_1_LIMIT - ((int) cpp));
+
+  *cpp++ = (char *)argcount;            /* the first value at cpp is argc */
+  cp2 = argbuf;
+  for (i = 0; i < argcount; i++) {      /* copy each argument and set argv */
+    *cpp++ = cp;
+    strcpy(cp, cp2);
+    cp += strlen(cp) + 1;
+    cp2 += strlen(cp2) + 1;
+  }
+  free(argbuf);
+  *cpp++ = NULL;                        /* the last argv is a NULL pointer */
+  *cpp++ = NULL;                        /* a NULL pointer for an empty envp */
+
+  return SUCCESS;
```

Comment:
Suppose proc with pid 2 is calling this to load a program on proc with pid 3.

when you start, r1 is set to pid 2's page table. 

then above, you set LoadProgram to use  `proc`'s page table, so pid 3 pt. 

when proc 2 continues on the return of this call, it'll expect to still have it's own page table. So you'll need to reset the page table back to the original state before returning. 

If you think that this is a flaw in the template code -- well, i agree :D.